[Drop-In UI] Expose public APIs via NavigationViewApi

tomaszrybakiewicz opened the pull request:
Closes [#1911](https://github.com/mapbox/navigation-sdks/issues/1911)

### Description

Changes:
- Introduced new`NavigationViewApi` methods:  `startFreeDrive`, `startRoutePreview`, `startNavigation`, `startArrival` and `endNavigation`.
- Refactored `NavigationViewApi` into abstract class and hid its implementation in `MapboxNavigationViewApi`
- Updated _libnavui-dropin_  module Metalava API file

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+    override fun startFreeDrive() {
+        store.dispatch(RoutesAction.SetRoutes(emptyList()))
+        store.dispatch(RoutePreviewAction.Ready(emptyList()))
+        store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
```
We should also set the destination to null

abhishek1508 replied:
Let's change this function to simply transition to `FreeDrive`

```
override fun startFreeDrive() {
    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

Move the other lines to `NavigationViewApi.startFreeDrive()` as an extension.

```
NavigationViewApi.startFreeDrive() {
    store.dispatch(DestinationAction.SetDestination(null))
    store.dispatch(RoutesAction.SetRoutes(emptyList()))
    store.dispatch(RoutePreviewAction.Ready(emptyList()))
    startFreeDrive()
}
```

tomaszrybakiewicz replied:
> Let's change this function to simply transition to FreeDrive
> ```
> override fun startFreeDrive() {
>     store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
> }
> ```

Isn't calling the proposed `startFreeDrive()` method to simply transition to the `FreeDrive` risk putting NavigationView in a corrupted state?
My understanding is that NavigationView should:
- enter **FreeDrive** when there is no `Destination` set and both `previewRoutes` and MapboxNavigation `routes` are empty
- enter **DestinationPreview** state when `Destination` is set and both `previewRoutes` and MapboxNavigation `routes` are empty
- enter **RoutePreview** state when `Destination` is set, `previewRoutes` is not empty and MapboxNavigation `routes` is empty
- enter **ActiveNavigation** state when `Destination` is set, `previewRoutes` and MapboxNavigation `routes` are not empty
- enter **Arrival** state when `Destination` is set, `previewRoutes` and MapboxNavigation `routes` are not empty, and `ArrivalObserver.onFinalDestinationArrival()` is called

<img width="931" alt="Screen Shot 2022-08-16 at 10 00 03 AM" src="https://user-images.githubusercontent.com/2678039/184899543-fd0f175e-6d8d-4774-9fe5-cf25e70cce99.png">

> Move the other lines to NavigationViewApi.startFreeDrive() as an extension.
> 
> ```
> NavigationViewApi.startFreeDrive() {
>     store.dispatch(DestinationAction.SetDestination(null))
>     store.dispatch(RoutesAction.SetRoutes(emptyList()))
>     store.dispatch(RoutePreviewAction.Ready(emptyList()))
>     startFreeDrive()
> }
> ```

You can't do that for a couple of reasons.
1. Proposed extension function doesn't have access to the `store`.
2. _If a class has a member function, and an extension function is defined which has the same receiver type, the same name, and is applicable to given arguments, the member always wins._ ([source](https://kotlinlang.org/docs/extensions.html#extensions-are-resolved-statically))

abhishek1508 replied:
The idea is to basically replicate what you are doing for `startRoutePReview` and `startActiveNavigation`.  

**Inside NavigationViewApi**
```
NavigationViewApi.startFreeDrive() {
    setDestination(null)
    setPreviewRoutes(empty)
    setRoutes(empty)
    startFreeDrive()
}
```

**Inside MapboxNavigationViewApi**
```
fun startFreeDrive() {
    // check destination is null
    // check route preview is null
    // check routes is null

    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

tomaszrybakiewicz replied:
```kotlin
fun MapboxNavigationViewApi.startFreeDrive() {
    // check destination is null
    // check route preview is null
    // check routes is null

    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

What's the benefit of having a method (^) that will always fail if the below methods are not called first?
```kotlin
setDestination(null)
setPreviewRoutes(empty)
setRoutes(empty)
```

Isn't it NavigationView responsibility to correctly update its internal state before moving to FreeDrive state?

```kotlin
fun MapboxNavigationViewApi.startFreeDrive() {
  setDestination(null)
  setPreviewRoutes(empty)
  setRoutes(empty)
  store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+    private val store: Store
+) : NavigationViewApi() {
+
+    override fun fetchRoutes(options: RouteOptions) {
+        store.dispatch(RoutePreviewAction.FetchOptions(options))
+    }
+
+    override fun fetchRoutes(points: List<Point>) {
+        store.dispatch(RoutePreviewAction.FetchPoints(points))
+    }
```
Based on https://github.com/mapbox/navigation-sdks/issues/1911#issuecomment-1214332643 we should remove the above `fetchRoutes` function.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+
+    @Throws(IllegalStateException::class)
+    override fun startRoutePreview() {
+        checkDestination()
```
We should not check for the existence of destination here. Doing this we are forcing end users to go through `DestinationPreview` state even if they don't want to.

tomaszrybakiewicz replied:
Ok. 
QQ: How do you get destination cooridnates from the `NavigationRoute`? Is this correct?

```kotlin
val destination: Point? = routes.first().routeOptions.coordinatesList().lastOrNull()
```

abhishek1508 replied:
That or we could also do `routes.first().directionsResponse.waypoints()?.lastOrNull()`.
@LukasPaczos Can you confirm is there a preference of one over the other?

LukasPaczos replied:
There are 3 ways to represent the destination:
- `routeOptions.coordinatesList().lastOrNull()` which is the requested destination
- `directionsResponse.waypoints()?.lastOrNull()` which is the requested destination snapped to a road network (so it's going to snapped to the road accessible closest to wherever the POI may be)
- `routeOptions.waypointTargetsList().lastOrNull()` which can be an additional representation of a destination that's presumably inaccessible from a road network or shouldn't be guide to directly (like a drop-off point).

Depending on on a scenario, either of these can be used. Additionally, there could be `directionsResponse.waypoints()` injected by the backend for multi-leg routes (charging stations for electric vehicles).

To unblock the PR, I think we should start with using `routes.first().directionsResponse.waypoints()?.lastOrNull()`, as that's guaranteed to be correct, however, it won't allow us to put a destination pin on the POI, the pin will be placed on the road. After that we should cut a ticket to determine what would we actually want to mark with the destination pin and build a short algorithm and data structure that can support it. We'll need a `Destination` type that holds both the routable point (snapped to the road network) and a destination point (which can be used to visualize the target). In this type of a data structure I think that the destination point should be determined like this: `routeOptions.waypointTargetsList().last` (if exists) -> `routeOptions.coordinatesList().last` (if exists) -> `directionsResponse.waypoints().last`.

Zayankovsky replied:
In 1TAP we use `routeOptions.coordinatesList().last()`. I see that Drop-In uses `directionsResponse.waypoints()?.lastOrNull()?.location()` and it doesn't look good in my opinion. As you can see in the video, I tapped on the POI, but the pin was placed in a different location. We already have grey circles denoting the waypoints, so I think we can use destination markers for something else. 

https://user-images.githubusercontent.com/2395284/187981836-240fc55a-2d80-4e2e-9e12-96402fc0624c.mp4

Zayankovsky replied:
Maybe `startRoutePreview` should have a parameter for specifying destination.

tomaszrybakiewicz replied:
@Zayankovsky, Could you open a new issue with your proposal?

Zayankovsky replied:
Added a comment here: https://github.com/mapbox/navigation-sdks/issues/1952#issuecomment-1234643487.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+        store.dispatch(NavigationStateAction.Update(NavigationState.Arrival))
+    }
+
+    override fun endNavigation() {
+        store.dispatch(endNavigationAction())
+    }
```
Based on my discussion with @LukasPaczos he gave a good suggestion in that we don't need this function, since `endNavigation` and `startFreeDrive` are doing the same thing. So in case, someone wants to end navigation they can just invoke `startFreeDrive`.

Let's remove this function.

abhishek1508 commented:
I missed it in the original ticket, but let's also add another function `startDestinationPreview(destination: Point)`

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+        store.dispatch(DestinationAction.SetDestination(Destination(point)))
+    }
+
+    override fun enableTripSession() {
+        store.dispatch(TripSessionStarterAction.EnableTripSession)
+    }
+
+    override fun enableReplaySession() {
+        store.dispatch(TripSessionStarterAction.EnableReplayTripSession)
+    }
```
This function should never need to be called by the developer in production, whether and when the trip session is started should be controlled by the use-cases and the state machine in the Drop-In UI itself.

I understand that these are exposed only for debug purposes anyway, so I'd expect something more along the lines of:
```
fun routeReplayEnabled(boolean)
```
or
```
fun debugModeEnabled(boolean, flags)
```
where one of the flags is `ROUTE_REPLAY`.

Then, whether a regular trip session should be started, or the replay one, or switched between them, and all the other integration details would be handled by the Drop-In UI.

tomaszrybakiewicz replied:
Ok. Which method would you like me to implement?

kmadsen replied:
Replay state is tied to the trip session state at the moment..

so `routeReplayEnabled(false)` would be ambiguous (is the trip session enabled or disabled?).

If we assume routeReplayEnabled(false) keeps the trip session enabled it would be the same as `enableTripSession`

LukasPaczos replied:
My concern is why would a developer interacting with Drop-In UI need to know if the trip session is enabled or disabled, this should be controlled internally by the Drop-In UI, shouldn't it?
```
fun debugModeEnabled(boolean, flags)
```
proposal assumes that Drop-In starts replaying when appropriate (when we're in active navigation, and maybe simulate the first point in preview), developer only sets whether they want that or not.

kmadsen replied:
Our api is still driven by how things work to some extent. We have to pick layers, of where to put the abstraction conversions. Are we saying that the `DropInUi_api` layer should be unaware of trip sessions? 

If we change the abstraction of what this function actually does it requires new logic.

If we want to change `enableReplaySession ` to include a `routeReplayEnabled(false)` action, there should be a [TripSessionStarterAction](https://github.com/mapbox/mapbox-navigation-android/blob/da3115462b7e081ea25676e495a6f92adb785c69/libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/tripsession/TripSessionStarterAction.kt#L8-L24) that includes something like a `DisableRouteReplay` action with a definition for what is actually going to happen. I'm just calling out that it is not implemented.

kmadsen replied:
But now it is :)

``` kotlin
    override fun routeReplayEnabled(enabled: Boolean) {
        if (enabled) {
            store.dispatch(TripSessionStarterAction.EnableReplayTripSession)
        } else {
            store.dispatch(TripSessionStarterAction.EnableTripSession)
        }
    }
```

kmadsen replied:
But I also, don't find it a big deal so will let other opinions weigh in to decide

kmadsen replied:
This is another example of logic existing in the API Layer. If this is what we want, I also think the `TripSessionStarterAction` should eventually be updated to have a `RouteReplayEnabled(boolean)` action. That way the logic layer can have tests

LukasPaczos reviewed (commented):
A couple of my thoughts from an earlier discussion with @abhishek1508 - in my opinion, Drop-In should be tailored towards handling encapsulated use-cases rather than standalone inputs. The granular controls of multiple functions that result in a certain state/side-effect is how our lower levels components work and Drop-In should build a level higher than that. This means that functions like `setRoutes(routes)` and `setDestination(point)` should be replaced with:
- `startDestinationPreview(point)`
- `startRoutesPreview(routes)`
- `startNavigation()` (that carries over from previous state)
- `startNavigation(point, routes)` (that starts from scratch)

and so on. Whatever state management is needed, for example `startNavigation(point, routes)` calling private `setRoutes(routes)` and `setDestination(point)` functions, should happen under the hood and not be required from the developer.

Building on top of that, each state has a dependency that could be provided directly or carried over from the previous state, that's why I'm thinking that instead of throwing exceptions when state requirements are not met, we should allow developers handle that gracefully.

For example:
```
fun startNavigation(point: Point? = null, routes: List<NavigationRoute>? = null, callback: StateChangeCallback? = null)

fun interface StateChangeCallback {
    fun onResult(result: Expected<StateChangeError, StateChangeSuccess>)
}

class StateChangeSuccess internal constructor()
class StateChangeError internal constructor(val type: Int) {
    companion object {
        const val TYPE_ROUTE_MISSING: Int = 0
    }
}
```

This is something we can discuss separately though.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewApi.kt:
```diff
@@ -73,36 +61,115 @@ class NavigationViewApi internal constructor(
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+@Throws(IllegalArgumentException::class)
+fun NavigationViewApi.startRoutePreview(destination: Point, routes: List<NavigationRoute>) {
```
Do we need `destination` as an input? We can deduce it internally and set it to the last point from `routes`.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewApi.kt:
```diff
@@ -73,36 +61,115 @@ class NavigationViewApi internal constructor(
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+@Throws(IllegalArgumentException::class)
+fun NavigationViewApi.startNavigation(destination: Point, routes: List<NavigationRoute>) {
```
Do we need destination as an input? We can deduce it internally and set it to the last point from `routes`.

tomaszrybakiewicz commented:
@LukasPaczos 
> Building on top of that, each state has a dependency that could be provided directly or carried over from the previous state, that's why I'm thinking that instead of throwing exceptions when state requirements are not met, we should allow developers handle that gracefully.
> 
> For example:
> 
> ```
> fun startNavigation(point: Point? = null, routes: List<NavigationRoute>? = null, callback: StateChangeCallback? = null)
> 
> fun interface StateChangeCallback {
>     fun onResult(result: Expected<StateChangeError, StateChangeSuccess>)
> }
> 
> class StateChangeSuccess internal constructor()
> class StateChangeError internal constructor(val type: Int) {
>     companion object {
>         const val TYPE_ROUTE_MISSING: Int = 0
>     }
> }
> ```
> 
> This is something we can discuss separately though.

@LukasPaczos Why can't we use exceptions? 
We explicitly document them in the KDoc and annotate each method with `@Throw` annotation. Kotlin already offers ways for developers to handle them gracefully via either `try / catch` expression or `runCatching(block: () -> R): Result<R> ` function that captures `block` exceptions and returns them as `Result` instance.

If we want to use the result object (Result or Expected) instead of exceptions, why do we need callback here? This method is synchronous. Can we just return the result object?

Furthermore, what's the benefit of having one method with default args 
  ```kotlin
  fun startNavigation(point: Point? = null, routes: List<NavigationRoute>? = null, callback: StateChangeCallback? = null))
  ```
vs. two methods
 ```kotlin
  fun startNavigation(): Result<Unit> // enter new state and carry over destination and routes data from previous state
  fun startNavigation(point: Point, routes: List<NavigationRoute>): Result<Unit> // enter new state with new destination and routes data
  ```
or 
```kotlin
  fun startNavigation(): Result<Unit> // enter new state and carry over destination and routes data from previous state
  fun startNavigation(routes: List<NavigationRoute>): Result<Unit> // enter new state with new routes data and destination point extracted form NavigationRoute last waypoint.
```

I am afraid that the "default args" variant will allow misuse of the API by permitting the passing of incorrect arguments:
```kotlin
navigationView.api.startNavigation(null, myRoutes)  // destination should never be null
```

tomaszrybakiewicz commented:
@LukasPaczos @abhishek1508
Looks like a lot of changes requested in this PR review comments significantly diverge from the [original ticket requirements](https://github.com/mapbox/navigation-sdks/issues/1911). Therefore, I propose moving any further discussion back to the [https://github.com/mapbox/navigation-sdks/issues/1911](https://github.com/mapbox/navigation-sdks/issues/1911) ticket.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,105 @@
+    override fun startFreeDrive() {
+        store.dispatch(RoutesAction.SetRoutes(emptyList()))
+        store.dispatch(RoutePreviewAction.Ready(emptyList()))
+        store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
```
Should this be needed? It seems the `NavigationStateController` should be able to figure out that it is in `FreeDrive` when there are no routes and the trip session is started.

Otherwise, this `startFreeDrive` appears to be a required function in order to put the `NavigationView` into free drive 🤔

tomaszrybakiewicz replied:
🤷 
remember [this](https://github.com/mapbox/mapbox-navigation-android/blob/e4fab080d536984928e60cc7f4c99ab6526fbf09/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/navigation/NavigationStateComponent.kt)?

abhishek1508 replied:
> Should this be needed? It seems the `NavigationStateController` should be able to figure out that it is in `FreeDrive` when there are no routes and the trip session is started.
> 
> Otherwise, this `startFreeDrive` appears to be a required function in order to put the `NavigationView` into free drive 🤔 

@kmadsen If an end user has a different use case and they want Drop-in to enter free drive state via different UI control, how would they do that if we don't have this function? Moreover, what you are saying is true in the case of the Core SDK. In Drop-In it's not sufficient to just set routes to null with trip session enabled to tell the user you are in free drive. You would need to update the `NavigationState` to free drive to control the UI. So for Drop-In, if you were to press on some custom external button in active navigation mode you need to tell Drop-In to set routes to null and transition to free drive. That's where this API comes in.

kmadsen replied:
I'm mostly skeptic of having _any logic_ inside this api layer.

Because where are the unit tests? Sure the function `startFreeDrive` should exist. But it seems it should be in a logic layer. For example, why isn't there a `StartFreeDriveAction`?

abhishek1508 replied:
I would like to have the logic outside API as well. Hence, I put up this [comment](https://github.com/mapbox/mapbox-navigation-android/pull/6174#discussion_r950639061) as well.

I would not block this PR though as it introduces the necessary public API(s) and rest all is implementation details which can be changed later.

tomaszrybakiewicz commented:
@abhishek1508 @LukasPaczos I've updated the `NavigationViewApi` implementation to match the latest spec. 
This one is ready for the next round of reviews.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,157 @@
+        store.dispatch(NavigationStateAction.Update(NavigationState.DestinationPreview))
+    }
+
+    override fun startRoutePreview(): Expected<Throwable, Unit> =
```
I don't have a strong opinion on `Expected<Throwable, Unit>` vs `Expected<SomeType, SomeTypeError>`. `Expected<Throwable, Unit>` does the job in a sense it returns appropriate error message when required.

However, I don't remember if as a team we made a decision to always return some dataType as a part of `Expected`. @LukasPaczos can you pitch in here?

LukasPaczos replied:
I would consider a full error type to give developers a chance to recover from a fault state.

If we return just a throwable, there's no way for a developer to recover in the runtime, they would have to treat the value as an unknown error and reset the whole experience back to square one. A more explicit type offers us a chance of providing error codes (specific for each of the state-mutating functions), for example, `ERROR_ROUTE_MISSING` for `startNavigation()` that would let the app check whether it has the route reference somewhere, set it and recover, or only in the worst case reset the whole experience. An explicit type also allows us to provide more information about the error that developers could push to their logs if needed, a string message isn't always enough.

That said, if we start with just `Throwable`, we can possibly extend it later if needed without it being a breaking chance. In any case, could we cut a ticket to explore this separately? I don't want this to block the PR.

abhishek1508 commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt:
```diff
@@ -0,0 +1,157 @@
+                "destination not found in a given route"
+            }
+
+            setDestination(point)
+            setPreviewRoutes(routes)
+            startRoutePreview()
```
🤔 If this and similarly others below can be extracted into `ThunkActions`? As far as I understand, that's why we introduced it to handle a group of actions.
@tomaszrybakiewicz

abhishek1508 replied:
@tomaszrybakiewicz when you get a chance can you pitch in here?

tomaszrybakiewicz replied:
Of course, we can introduce `ThunkActions` for these groups of actions. Technically, every method in this class could be turned into a ThunkAction. However, I haven't found other use cases where those actions would also be dispatched, therefore I left that refactoring till later.

LukasPaczos commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewApi.kt:
```diff
@@ -1,108 +1,109 @@
-    fun setDestination(point: Point) {
-        store.dispatch(DestinationAction.SetDestination(Destination(point)))
-    }
+    abstract fun startNavigation(): Expected<Throwable, Unit>
```
```suggestion
    abstract fun startActiveGuidance(): Expected<Throwable, Unit>
```
to be consistent with our `Free Drive` / `Active Guidance` terminology

abhishek1508 commented:
Thanks for pitching everyone. I have created this ticket > https://github.com/mapbox/navigation-sdks/issues/1952 that tracks further improvements on top of this PR. Let's add a changelog @tomaszrybakiewicz 

✅ the PR so you can address simple remaining changes and merge it today.

cc @LukasPaczos

tomaszrybakiewicz commented:
@LukasPaczos I've updated NavigationViewApi error handling as you suggested. I introduced the `NavigationViewApiError` class and updated the `NavigationViewApi` implementation:
```kotlin
/**
 * Errors returned by the NavigationApi.
 */
sealed class NavigationViewApiError(message: String) : Throwable(message) {
    /**
     * Error returned when the Destination hasn't been set yet.
     */
    object MissingDestinationInfo : NavigationViewApiError("Destination cannot be empty.")

    /**
     * Error returned when the Preview Routes list hasn't been set yet.
     */
    object MissingPreviewRoutesInfo : NavigationViewApiError("Preview Routes cannot be empty.")

    /**
     * Error returned when the Routes list hasn't been set yet.
     */
    object MissingRoutesInfo : NavigationViewApiError("Routes cannot be empty.")

    /**
     * Error returned when given PreviewRoute or Route list is empty.
     */
    object InvalidRoutesInfo : NavigationViewApiError("Routes cannot be empty.")

    /**
     * Error returned when given PreviewRoute or Route list is missing [DirectionsWaypoint]
     * information that is needed to determine Destination coordinates.
     */
    object IncompleteRoutesInfo :
        NavigationViewApiError("Missing destination info in a given route.")
}
```
Please take a look when you have a moment. I will wait for your approval before merging.

cc: @abhishek1508

LukasPaczos reviewed (commented):
The error handling proposed here is a good start :+1: We'll need to consider two additional topics:
- Should the error types be more explicit and directly document which data points (via which method) should be provided first?
- A sealed class is an enumeration, and if used inside of a `when` _expression_, all branches need to be covered, or an `else` fallback has to be present. This means that adding a type to a sealed class is a binary incompatible change, because the compiler will require the caller to cover all the branches if they didn't use an `else` fallback. We could consider using an `IntDef`-like non-restrictive approach instead.

Let's merge as is to unblock the downstream test integration but especially the latter topic has to be considered a stable release blocker.

Zayankovsky commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewApi.kt:
```diff
@@ -1,108 +1,145 @@
+     * [NavigationViewListener.onActiveNavigation] will be called once [NavigationView] enters
+     * Route Preview state.
+     *
+     * Fails with an error when either Destination or Preview Routes has not been set.
```
What is the reason for this limitation? In 1TAP we need an ability to start active guidance without any routes. The use case is replaying history files, that don't have any routes in case of a free drive recording. For that reason we continue using `SharedApp.store.dispatch(NavigationStateAction.Update(NavigationState.ActiveNavigation))`.

Zayankovsky commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewApi.kt:
```diff
@@ -1,108 +1,145 @@
+     * [NavigationViewListener.onArrival] will be called once [NavigationView] enters
+     * Arrival state.
+     *
+     * Fails with an error when either Destination or Routes has not been set.
```
As with `startActiveGuidance` this limitation also doesn't let replace `SharedApp.store.dispatch(NavigationStateAction.Update(NavigationState.Arrival))` with `startArrival` in 1TAP. The use case where this doesn't work for us is showing arrival feedback for the trip that happened a while ago. We save metadata for the trip that is necessary for reporting the feedback later, but we can't afford saving the whole routes.
